taskcat/_cfn/stack.py

```diff
@@ -3,44 +3,57 @@
 import random
 import re
 import string
-from datetime import datetime
+from datetime import datetime, timedelta
 from pathlib import Path
-from typing import Any, Dict, List, Optional
+from threading import Timer
+from typing import Callable, List, Optional, Tuple
 from uuid import UUID, uuid4
 
+import boto3
+
 from taskcat._cfn.template import Template
-from taskcat._client_factory import ClientFactory
+from taskcat._common_utils import pascal_to_snake, s3_url_maker
+from taskcat._config import S3Bucket
+from taskcat._config_types import AWSRegionObject
 
 GENERIC_ERROR_PATTERNS = [
     r"(The following resource\(s\) failed to create: )",
     r"(^Resource creation cancelled$)",
 ]
 
 
+def criteria_matches(criteria: dict, instance):
+    # fail if criteria includes an invalid property
+    for k in criteria:
+        if k not in instance.__dict__:
+            raise ValueError(f"{k} is not a valid property of {type(instance)}")
+    for k, v in criteria.items():
+        # matching is AND for multiple criteria, so as soon as one fails,
+        # it's not a match
+        if getattr(instance, k) != v:
+            return False
+    return True
+
+
 class StackStatus:
-    CREATE_IN_PROGRESS = ["CREATE_IN_PROGRESS"]
-    CREATE_COMPLETE = ["CREATE_COMPLETE"]
-    CREATE_FAILED = [
+    COMPLETE = ["CREATE_COMPLETE", "UPDATE_COMPLETE", "DELETE_COMPLETE"]
+    IN_PROGRESS = [
+        "CREATE_IN_PROGRESS",
+        "DELETE_IN_PROGRESS",
+        "UPDATE_IN_PROGRESS",
+        "UPDATE_COMPLETE_CLEANUP_IN_PROGRESS",
+    ]
+    FAILED = [
```

Comment:
Why was this reduced to COMPLETE/IN_PROGRESS/FAILED? 

I can see many scenarios where knowing a DELETE_FAILED occured (verus CREATE_FAILED) is useful.

Reply:
you will know specific status (and can query it):

```
s.stacks[0].status
Out[55]: 'CREATE_FAILED'
```

These are convenience groupings for the cases when you only want to know the general state. For example, line 390, where there is no point in continuing to poll `describe_stack` for a stack that's not in an `IN_PROGRESS` state